#! /bin/sh /usr/share/dpatch/dpatch-run
## runloop-backchannel-eof-spin.dpatch by Samuel Thibault <samuel.thibault@ens-lyon.org>
##
## DP: Fix backend runloop spin on backchannel EOF (select() returns "ready for
## DP: read" on EOF). This completely broke printing with e. g. HPJetDirect.
## DP: (http://bugs.debian.org/489045)
@DPATCH@
diff -urNad trunk~/backend/runloop.c trunk/backend/runloop.c
--- trunk~/backend/runloop.c	2008-06-16 19:41:11.000000000 +0200
+++ trunk/backend/runloop.c	2008-11-13 12:11:52.000000000 +0100
@@ -161,6 +161,7 @@
 #if defined(HAVE_SIGACTION) && !defined(HAVE_SIGSET)
   struct sigaction action;		/* Actions for POSIX signals */
 #endif /* HAVE_SIGACTION && !HAVE_SIGSET */
+  int device_input = 1;
 
 
   fprintf(stderr,
@@ -210,7 +211,7 @@
     FD_ZERO(&input);
     if (!print_bytes)
       FD_SET(print_fd, &input);
-    if (use_bc)
+    if (use_bc && device_input)
       FD_SET(device_fd, &input);
     if (!print_bytes && side_cb)
       FD_SET(CUPS_SC_FD, &input);
@@ -273,6 +274,13 @@
 	        CUPS_LLCAST bc_bytes);
         cupsBackChannelWrite(bc_buffer, bc_bytes, 1.0);
       }
+      else if (bc_bytes == 0)
+      {
+       /*
+        * End of file, stop trying to read from device_fd
+	*/
+        device_input = 0;
+      }
     }
 
    /*
